### PR TITLE
PP-745: Fix for single node job, string type resource encoding.

### PR DESCRIPTION
**Pull request by dilip-krishnan**
#### Issue-ID

* **[PP-745](https://pbspro.atlassian.net/browse/PP-745)**

#### Problem description
* Single node jobs don't quote JSON formatted strings.

#### Cause / Analysis
* While accumulating resources for a multinode job string type  resources are merge if
they are in json format  and are set within single quotes. But in case of single node job such
accumulation process was not required, due to which string resources having json string were
not quoted. This causes an inconsistent of resource usage log in accounting log. 

#### Solution description
* In case of single node job, in encode_resc_used check for json string resources and set them within single quotes.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
+import time
+from tests.functional import *
+
+
+class TestPbsPP745(TestFunctional):
```
Comment (Klovely) [inline test/tests/functional/pbs_resc_used_single_node.py]: We can give a more meaningful class name instead of TestPbs745.

**Comment by prakashcv13** [comment]
@dilip-krishnan, We would also need a external interface design document for these changes.

**Comment by dilip-krishnan** [comment]
Hi @prakashcv13  & @Klovely ,
     I have addressed your comments. I am working on the EDD and running the pbs_accumulate.. test case.
Please review the latest changes.

Warm regards
Dilip

**Comment by Klovely** [comment]
Changes looks good to me!

**Comment by dilip-krishnan** [comment]
Hi @PBSPro/pbspro-maintainers,
         This PR is ready for merge please review the changes.

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    rsc_list = ['foo_str', 'foo_f', 'foo_i', 'foo_str2', 'foo_str3']
+
+    def tearDown(self):
+        self.du.set_pbs_config(confs={'PBS_SERVER': self.server.hostname})
+        TestFunctional.tearDown(self)
+        for r in self.rsc_list:
+            try:
+                self.server.manager(MGR_CMD_DELETE, RSC, id=r, runas=ROOT_USER)
+            except:
+                pass
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: You don't need to remove resources in tearDown().  revert_to_defaults() does that.

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        self.server.restart()
+        self.mom.restart()
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        for r in self.rsc_list:
+            try:
+                self.server.manager(MGR_CMD_DELETE, RSC, id=r, runas=ROOT_USER)
+            except:
+                pass
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: These resources won't exist.  revert_to_defaults() is called in TestFunctional.setUp().  It removes all custom resources.

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+        self.server.restart()
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: Why restart the server?

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+        self.momA = self.moms.values()[0]
+        self.momA.restart()
+        self.momA.delete_vnode_defs()
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: Why restart the mom before deleting the vnode def file?

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        self.assertEqual(rc, 0)
+        self.server.expect(NODE, {'state': 'free'}, id=self.hostA)
+
+        a = {'job_history_enable': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        # Next set some custom resources via qmgr -c 'create resource'
+        attr = {}
+        attr['type'] = 'string'
+        attr['flag'] = 'h'
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: Suggestion: Combine these lines into the initialization of attr.  This means attr = {'type': 'string', 'flag': 'h'}

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        r = 'foo_str2'
+        rc = self.server.manager(
+            MGR_CMD_CREATE, RSC, attr, id=r, runas=ROOT_USER, logerr=False)
+        self.assertEqual(rc, 0)
+
+        attr['type'] = 'string'
+        attr['flag'] = 'h'
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: No need to reset attr.  It's still set from above.

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        r = 'foo_str3'
+        rc = self.server.manager(
+            MGR_CMD_CREATE, RSC, attr, id=r, runas=ROOT_USER, logerr=False)
+        self.assertEqual(rc, 0)
+
+        attr['type'] = 'string_array'
+        attr['flag'] = 'h'
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: The flag hasn't changed from above.  No need to reset it.

File: test/tests/functional/pbs_resc_used_single_node.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under a
+# commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+import time
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: Question: Why break this test out from the test plan that tests single and multi node accumulation?  It just proliferates the number of test plans.

Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Hi @bhroam,
     On my dev Centos 7 machine, i noticed that resources created were not getting cleaned in ptl by revert_to_defaults. Due to which the test case failed if ran multiple time  due resource  Duplicate resource entry already exist error. Also Multi node test case does a minimum 3 mom requirement test in setup and many other multi-node related setup. Hence too much refactoring was required in the existing test to just make single node test case work. Therefore it put in a separate file.

Regards

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 2 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: I noticed that revert_to_defaults was not able to remove resource created by ptl on my centos 7 dev machine. I had to manually delete and need to restart mom , server services. To restore to defaults.This behaviour is observed on both master and PP-745.

Hence I added additional step, just to make sure to delete resources get removed before and after the test case ran

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 3 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Same reason as above.

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 4 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Same reason as above.

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 5 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Same as above.

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 6 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Done.

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 7 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Done

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 8 above)
Comment (dilip-krishnan) [inline test/tests/functional/pbs_resc_used_single_node.py]: Done

File: test/tests/functional/pbs_resc_used_single_node.py (hunk 9 above)
Comment (bhroam) [inline test/tests/functional/pbs_resc_used_single_node.py]: Strange, I've seen PTL remove resources before.  If it isn't removing the resources, then that's a PTL bug.  Please file it.  As for multiple files, I understand your reasoning.  This isn't the first time tests have been split between single/multi-node tests.  I'd rather see them in the same file, but you are correct about the refactoring needed.  I'm fine with it.

**Comment by bhroam** [comment]
@dilip-krishnan please squash and rebase for preparation for the final merge.

**Comment by dilip-krishnan** [comment]
Hi @bhroam ,
    I have rebased and sqaushed the PR to single commit.

Regards 
Dilip

**Comment by bhroam** [comment]
Looks good, merged.
